File: lib/extensions/network.dart
import 'package:bitcoin_ui/bitcoin_ui.dart';
import 'package:danawallet/constants.dart';
import 'package:danawallet/generated/rust/api/structs/network.dart';
import 'package:danawallet/global_functions.dart';
import 'package:flutter/services.dart';

extension NetworkExtension on ApiNetwork {
  String get defaultBlindbitUrl {
    switch (this) {
      case ApiNetwork.mainnet:
        if (isDevEnv && const String.fromEnvironment("MAINNET_URL") != "") {
          return const String.fromEnvironment("MAINNET_URL");
        } else {
          return defaultMainnet;
        }
      case ApiNetwork.testnet3:
      case ApiNetwork.testnet4:
        if (isDevEnv && const String.fromEnvironment("TESTNET_URL") != "") {
          return const String.fromEnvironment("TESTNET_URL");
        } else {
          return defaultTestnet;
        }
      case ApiNetwork.signet:
        if (isDevEnv && const String.fromEnvironment("SIGNET_URL") != "") {
          return const String.fromEnvironment("SIGNET_URL");
        } else {
          return defaultSignet;
        }
      case ApiNetwork.regtest:
        if (isDevEnv && const String.fromEnvironment("REGTEST_URL") != "") {
          return const String.fromEnvironment("REGTEST_URL");
        } else {
          return defaultRegtest;
        }
    }
  }

  Color get toColor {
    switch (this) {
      case ApiNetwork.mainnet:
        return Bitcoin.orange;
      case ApiNetwork.testnet3:
      case ApiNetwork.testnet4:
        return Bitcoin.green;
      case ApiNetwork.signet:
        return Bitcoin.purple;
      case ApiNetwork.regtest:
        return Bitcoin.blue;
    }
  }

  int get defaultBirthday {
    switch (this) {
      case ApiNetwork.mainnet:
        return defaultMainnetBirthday;
      case ApiNetwork.testnet3:
      case ApiNetwork.testnet4:
        return defaultTestnetBirthday;
      case ApiNetwork.signet:
        return defaultSignetBirthday;
      case ApiNetwork.regtest:
        return defaultRegtestBirthday;
    }
  }
}

File: rust/src/api/structs/network.rs
use serde::{Deserialize, Serialize};
use spdk_wallet::bitcoin;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ApiNetwork {
    Mainnet,
    Testnet3,
    Testnet4,
    Signet,
    Regtest,
}

impl From<bitcoin::Network> for ApiNetwork {
    fn from(value: bitcoin::Network) -> Self {
        match value {
            bitcoin::Network::Bitcoin => ApiNetwork::Mainnet,
            bitcoin::Network::Testnet => ApiNetwork::Testnet3,
            bitcoin::Network::Testnet4 => ApiNetwork::Testnet4,
            bitcoin::Network::Signet => ApiNetwork::Signet,
            bitcoin::Network::Regtest => ApiNetwork::Regtest,
        }
    }
}

impl From<ApiNetwork> for bitcoin::Network {
    fn from(value: ApiNetwork) -> Self {
        match value {
            ApiNetwork::Mainnet => bitcoin::Network::Bitcoin,
            ApiNetwork::Testnet3 => bitcoin::Network::Testnet,
            ApiNetwork::Testnet4 => bitcoin::Network::Testnet4,
            ApiNetwork::Signet => bitcoin::Network::Signet,
            ApiNetwork::Regtest => bitcoin::Network::Regtest,
        }
    }
}
